### PR TITLE
Fix #17393

### PR DESCRIPTION
**Pull request by pedro-hos**
update the code for the new implementation methods and tested on Keycloak 21.0.1

Closes: keycloak/keycloak#17393

### CODE REVIEW

**Review by vramik — CHANGES_REQUESTED** [review]
Thanks @pedro-hos for the PR. Overall changes lgtm. I have just few minor comments.

- Please keep the overriding methods in appropriate sections. Please see comments inside a classes like

```java
// UserQueryProvider methods
// UserLookupProvider methods
// CredentialInputValidator methods
// CredentialInputUpdater methods
```

- Could you please check indentation of your code and keep it in line with current code, e.g.
```java
        @Override
	public UserModel getUserById(RealmModel realm, String id) {
		StorageId storageId = new StorageId(id);
        String username = storageId.getExternalId();
        return getUserByUsername(realm, username);
	}
``` 
doesn't look good. 

- please make sure that commit message contains original issue reference, like `Fixes keycloak/keycloak#17393`.

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 1]
```diff
@@ -284,16 +181,96 @@ public void disableCredentialType(RealmModel realm, UserModel user, String crede
...
+            users.add(user);
+            if (users.size() >= maxResults) break;
+        }
+        return users.stream();
+	}
+
+	@Override
+	public Stream<UserModel> searchForUserStream(RealmModel realm, Map<String, String> params) {
+        return searchForUserStream(realm, params, 0, Integer.MAX_VALUE);
+	}
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:218]: This method has default implementation, this seems as redundant:
 
https://github.com/keycloak/keycloak/blob/9052ec2b02f0aace57c5ffb6fb472345a56b11cf/server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java#L228-L230

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 2]
```diff
@@ -284,16 +181,96 @@ public void disableCredentialType(RealmModel realm, UserModel user, String crede
...
+    public void close() {
 
-        return disableableTypes;
     }
+    
     @Override
-    public void close() {
-
+	public Stream<UserModel> searchForUserStream(RealmModel realm, String search) {
+    	return searchForUserStream(realm, search, 0, Integer.MAX_VALUE);
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:194]: This method has default implementation, this seems as redundant:

https://github.com/keycloak/keycloak/blob/main/server-spi/src/main/java/org/keycloak/storage/user/UserQueryProvider.java#L185-L187

**Comment by pedro-hos** [comment]
@vramik I've just updated with the requested changes. Thanks a lot

**Review by vramik — CHANGES_REQUESTED** [review]
Thanks a lot for the latest changes @pedro-hos, it looks much better now. 

I'd just suggest to use original indentation (4 spaces) to be able to see only real diff.

In current state everything is marked as changed.

**Comment by pedro-hos** [comment]
@vramik  done =D

File: user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java [hunk 3]
```diff
@@ -17,39 +17,38 @@
...
-        UserLookupProvider,
-        CredentialInputValidator,
-        CredentialInputUpdater
-{
+public class PropertyFileUserStorageProvider implements 
+        UserStorageProvider, 
+        UserLookupProvider, 
+        CredentialInputValidator, 
+        CredentialInputUpdater {
+
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java:51]: These changes seems to be unnecessary.

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 4]
```diff
@@ -125,79 +126,56 @@ public int getUsersCount(RealmModel realm) {
...
+            String username = (String) obj;
+            if (!username.contains(search))
+                continue;
+            if (i++ < firstResult)
+                continue;
+            UserModel user = getUserByUsername(realm, username);
             users.add(user);
-            if (users.size() >= maxResults) break;
+            if (users.size() >= maxResults)
+                break;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:145]: Also not necessary, please revert.

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 5]
```diff
@@ -125,79 +126,56 @@ public int getUsersCount(RealmModel realm) {
...
-    @Override
-    public List<UserModel> searchForUser(Map<String, String> params, RealmModel realm, int firstResult, int maxResults) {
+    public Stream<UserModel> searchForUserStream(RealmModel realm, Map<String, String> params, Integer firstResult,
+            Integer maxResults) {
         // only support searching by username
         String usernameSearchString = params.get("username");
-        if (usernameSearchString == null) return Collections.EMPTY_LIST;
-        return searchForUser(usernameSearchString, realm, firstResult, maxResults);
+        if (usernameSearchString == null)
+            return Stream.empty();
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:156]: please revert

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 6]
```diff
@@ -224,46 +202,47 @@ public UserModel addUser(RealmModel realm, String username) {
     @Override
     public boolean removeUser(RealmModel realm, UserModel user) {
         synchronized (properties) {
-            if (properties.remove(user.getUsername()) == null) return false;
+            if (properties.remove(user.getUsername()) == null)
+                return false;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:206]: please revert

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 7]
```diff
@@ -224,46 +202,47 @@ public UserModel addUser(RealmModel realm, String username) {
...
     public boolean supportsCredentialType(String credentialType) {
-        return credentialType.equals(CredentialModel.PASSWORD);
+        return credentialType.equals(PasswordCredentialModel.TYPE);
     }
 
     @Override
     public boolean isValid(RealmModel realm, UserModel user, CredentialInput input) {
-        if (!supportsCredentialType(input.getType()) || !(input instanceof UserCredentialModel)) return false;
+        if (!supportsCredentialType(input.getType()) || !(input instanceof UserCredentialModel))
+            return false;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:228]: please revert

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 8]
```diff
@@ -224,46 +202,47 @@ public UserModel addUser(RealmModel realm, String username) {
...
 
-        UserCredentialModel cred = (UserCredentialModel)input;
+        UserCredentialModel cred = (UserCredentialModel) input;
         String password = properties.getProperty(user.getUsername());
-        if (password == null || UNSET_PASSWORD.equals(password)) return false;
+        if (password == null || UNSET_PASSWORD.equals(password))
+            return false;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:233]: please revert

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 9]
```diff
@@ -224,46 +202,47 @@ public UserModel addUser(RealmModel realm, String username) {
...
     @Override
     public boolean updateCredential(RealmModel realm, UserModel user, CredentialInput input) {
-        if (!(input instanceof UserCredentialModel)) return false;
-        if (!input.getType().equals(CredentialModel.PASSWORD)) return false;
-        UserCredentialModel cred = (UserCredentialModel)input;
+        if (!(input instanceof UserCredentialModel))
+            return false;
+        if (!input.getType().equals(PasswordCredentialModel.TYPE))
+            return false;
+        UserCredentialModel cred = (UserCredentialModel) input;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:245]: please revert formatting changes (and keep `CredentialModel.PASSWORD` -> `PasswordCredentialModel.TYPE`)

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java [hunk 10]
```diff
@@ -273,7 +252,8 @@ public boolean updateCredential(RealmModel realm, UserModel user, CredentialInpu
 
     @Override
     public void disableCredentialType(RealmModel realm, UserModel user, String credentialType) {
-        if (!credentialType.equals(CredentialModel.PASSWORD)) return;
+        if (!credentialType.equals(PasswordCredentialModel.TYPE))
+            return;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:256]: please revert formatting changes (and keep `CredentialModel.PASSWORD` -> `PasswordCredentialModel.TYPE`)

File: user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java [hunk 11]
```diff
@@ -92,46 +91,49 @@ public SubjectCredentialManager credentialManager() {
...
         return password.equals(cred.getValue());
     }
 
     // CredentialInputUpdater methods
 
     @Override
     public boolean updateCredential(RealmModel realm, UserModel user, CredentialInput input) {
-        if (input.getType().equals(CredentialModel.PASSWORD)) throw new ReadOnlyException("user is read only for this update");
+        if (input.getType().equals(PasswordCredentialModel.TYPE))
+            throw new ReadOnlyException("user is read only for this update");
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java:136]: please revert formatting changes (and keep `CredentialModel.PASSWORD` -> `PasswordCredentialModel.TYPE`)

File: user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java [hunk 12]
```diff
@@ -92,46 +91,49 @@ public SubjectCredentialManager credentialManager() {
...
     public boolean supportsCredentialType(String credentialType) {
-        return credentialType.equals(CredentialModel.PASSWORD);
+        return credentialType.equals(PasswordCredentialModel.TYPE);
     }
 
     @Override
     public boolean isValid(RealmModel realm, UserModel user, CredentialInput input) {
-        if (!supportsCredentialType(input.getType()) || !(input instanceof UserCredentialModel)) return false;
+        if (!supportsCredentialType(input.getType()) || !(input instanceof UserCredentialModel))
+            return false;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java:122]: please revert

File: user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java [hunk 13]
```diff
@@ -92,46 +91,49 @@ public SubjectCredentialManager credentialManager() {
...
 
-        UserCredentialModel cred = (UserCredentialModel)input;
+        UserCredentialModel cred = (UserCredentialModel) input;
         String password = properties.getProperty(user.getUsername());
-        if (password == null) return false;
+        if (password == null)
+            return false;
```
Comment (vramik) [inline user-storage-simple/src/main/java/org/keycloak/quickstart/readonly/PropertyFileUserStorageProvider.java:127]: please revert

**Review by vramik — CHANGES_REQUESTED** [review]
Thank you very much @pedro-hos for the update. I've got only last comment regarding unnecessary formatting changes. Please see inline. 

After addressing it I think it's ready to be merged.

File: user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java (hunk 9 above)
Comment (pedro-hos) [reply user-storage-simple/src/main/java/org/keycloak/quickstart/writeable/PropertyFileUserStorageProvider.java:245]: Hi @vramik, one question, `CredentialModel.PASSWORD` is deprecated, should I keep it?

**Comment by pedro-hos** [comment]
@vramik thank you very much for the recommendations. Hope to fix all indentations now =D

File: user-storage-simple/README.md [hunk 14]
```diff
@@ -39,13 +39,30 @@ All you need to build this project is Java 8.0 (Java SDK 1.8) or later and Maven
...
+./kc.sh start-dev
+```
+
+Then type the follow maven command:
+
+   ````
+   mvn clean install
+   ````
+Finally copy the `.jar` file into `/providers` folder on <span>Keycloak</span>
+
```
Comment (vramik) [inline user-storage-simple/README.md:65]: @pedro-hos I've just realized hot deploying doesn't work with quarkus based dist. Therefore this section has to be updated in a way that first has to be jar copied and then server could be started. 

Sorry for late response. 

I'd even suggest to remove Wildfly part all together.

**Review by vramik — CHANGES_REQUESTED** [review]

File: user-storage-simple/README.md (hunk 14 above)
Comment (pedro-hos) [reply user-storage-simple/README.md:65]: I agree! Also, I am wondering if I need to remove the following as well:

~~~
Integration test of the Quickstart
----------------------------------

1. Make sure you have an Keycloak server running with an admin user in the `master` realm or use the provided docker image
2. You need to have Chrome browser installed and updated to the latest version.
3. Run `mvn test -Pkeycloak-remote`
~~~

Comment (vramik) [reply user-storage-simple/README.md]: It'd probably be good to update it as well, together with checking the test as well.  

I need to find out current plans regarding the tests. There is: https://github.com/keycloak/keycloak-quickstarts/blob/69556a965568e68fb85e125b497262b2d56df074/scripts/run-tests.sh#L85-L92

Let me check it and get back to you once I know more, thanks for the patience.

File: user-storage-simple/README.md [hunk 15]
```diff
@@ -60,13 +67,6 @@ admin console or on the registration pages of <span>Keycloak</span>, will be cre
 to the Users tab in the Admin Console and create a new user, you'll be able to see the provider in action.  You can also
 edit the file yourself to add the username/password pairs you want.
 
-Integration test of the Quickstart
-----------------------------------
-
-1. Make sure you have an Keycloak server running with an admin user in the `master` realm or use the provided docker image
-2. You need to have Chrome browser installed and updated to the latest version.
-3. Run `mvn test -Pkeycloak-remote`
-
```
Comment (vramik) [inline user-storage-simple/README.md:69]: @pedro-hos I've checked it and I can confirm these steps are still valid. Could you revert this change? 

I'd just add to the first point that keycloak has to be listening on `http://localhost:8180/auth`.

Then `createTestArchive` method could be removed from `ArquillianSimpleStorageTest`.

**Review by vramik — CHANGES_REQUESTED** [review]

File: user-storage-simple/README.md (hunk 15 above)
Comment (vramik) [reply user-storage-simple/README.md:69]: And on top of that if you could cherry-pick this [commit](https://github.com/keycloak/keycloak-quickstarts/commit/f9fd95a03c3a683399d8624a49a69c60e77843a3), it'd add the intergation test in GHA. Thanks

Comment (pedro-hos) [reply user-storage-simple/README.md]: @vramik I have finished the changes! Thanks

**Review by vramik — APPROVED** [review]
Tests results:
```
2023-03-09T14:58:24.4453307Z Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 15.883 sec - in org.keycloak.quickstart.ArquillianSimpleStorageTest
```

```
2023-03-09T14:59:13.7953299Z FAILED TESTS:
2023-03-09T14:59:13.7954107Z - app-springboot
2023-03-09T14:59:13.7954756Z - service-nodejs
```

Thank you @pedro-hos.

**Review by hmlnarik — APPROVED** [review]
Approving per @vramik 's approval. Thank you @pedro-hos !

**Comment by vmuzikar** [comment]
Just a silly question, was it intentional this PR went to `latest` branch and not `main`? AFAIK,  `latest` is overwritten with each release.

**Comment by vramik** [comment]
> was it intentional this PR went to `latest` branch and not `main`

no, it was mistake from my site, I didn't noticed it during review. I've cherry-picked it and sent to `main`: https://github.com/keycloak/keycloak-quickstarts/pull/386

**Comment by pedro-hos** [comment]
Thank you @vramik, I am sorry for the branch confusion. I will use main instead latest for other contributions =D. Thanks @vmuzikar and @hmlnarik

**Pull request merged**